src/main/java/org/jenkinsci/plugins/pipeline/milestone/MilestoneStep.java

```diff
@@ -93,10 +118,37 @@ public boolean isUnsafe() {
         return unsafe;
     }
 
+    @CheckForNull
+    public String getGroup() {
+        return group;
+    }
+
+    @CheckForNull
+    public MilestonePolicy getPolicy() {
+        return policy;
+    }
+
+    @Override
+    public String toString() {
+        return "MilestoneStep["
+                +" group="+group
+                +", ordinal="+ordinal
+                +", label="+label
+                +"]";
+    }
+
+    public ListBoxModel doFillPolicyItems() {
+        ListBoxModel r = new ListBoxModel();
+        for (MilestonePolicy policy : MilestonePolicy.values()) {
+            r.add(policy.name());
+        }
+        return r;
+    }
+
     @Extension
     public static final class DescriptorImpl extends AbstractStepDescriptorImpl {
 
-        private Map<String, Map<Integer, Milestone>> milestonesByOrdinalByJob;
+        private Map<String, Map<String, Milestone>> milestonesByGroupByJob;
```

Comment:
This is a backward incompatible change, which means in-progress builds using milestones will behave like if there were no milestones at all when Jenkins is restarted after the plugin upgrade (and possibly a deserialization exception). It needs some migration code.

Reply:
@amuniz do you have any examples of migration code? (maybe in other plugins) 
I believe I need to use readResolve but I have no idea how to implement it and test.

Reply:
Yes `readResolve` can be used in Jenkins plugin code including for this case, XStream-serialized XML files. Would normally be tested via `@LocalData`.